pkg/graph/resolvers/trb_request_document.go

```diff
@@ -0,0 +1,89 @@
+package resolvers
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/google/uuid"
+
+	"github.com/cmsgov/easi-app/pkg/appcontext"
+	"github.com/cmsgov/easi-app/pkg/graph/model"
+	"github.com/cmsgov/easi-app/pkg/models"
+	"github.com/cmsgov/easi-app/pkg/storage"
+	"github.com/cmsgov/easi-app/pkg/upload"
+)
+
+const fallbackExtension = ".unknown"
+
+// GetTRBRequestDocumentsByRequestID fetches all documents attached to the TRB request with the given ID.
+func GetTRBRequestDocumentsByRequestID(ctx context.Context, store *storage.Store, s3Client *upload.S3Client, requestID uuid.UUID) ([]*models.TRBRequestDocument, error) {
+	documents, err := store.GetTRBRequestDocumentsByRequestID(appcontext.ZLogger(ctx), requestID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, document := range documents {
+		presignedURL, err := s3Client.NewGetPresignedURL(document.S3Key)
```

Comment:
A couple things here - these calls to get tags and URLs from S3 are all blocking, and in a loop. You could use goroutines and a [WaitGroup](https://pkg.go.dev/sync#WaitGroup). [Here](https://medium.com/@greenraccoon23/multi-thread-for-loops-easily-and-safely-in-go-a2e915302f8b) is a good example.

Also, this URL will never change, right? Perhaps it could be stored once in the database, that way we won't need to make two requests per document, and only need to fetch the tag.

Reply:
From some testing, in both cases, a single error fetching metadata for any single document causes the whole GQL request to error; it happens if there's a separate resolver for the URL field, or if using an errgroup inside the existing `GetTRBRequestDocumentsByRequestID()` function. Additionally, the errgroup method's code is a _lot_ messier:

```go
func GetTRBRequestDocumentsByRequestID(ctx context.Context, store *storage.Store, s3Client *upload.S3Client, requestID uuid.UUID) ([]*models.TRBRequestDocument, error) {
	documents, err := store.GetTRBRequestDocumentsByRequestID(appcontext.ZLogger(ctx), requestID)
	if err != nil {
		return nil, err
	}

  // new code starts here
	g := new(errgroup.Group)

  // have to use sync.Map because the built-in maps aren't goroutine-safe. Alternatively, we could use mutexes or something, but that's even more complicated.
	presignedURLs := sync.Map{}
	documentStatuses := sync.Map{}

	for _, document := range documents {
		document := *document
		g.Go(func() error {
			presignedURL, err := s3Client.NewGetPresignedURL(document.S3Key)
			if err != nil {
				return err
			}
			presignedURLs.Store(document.ID, presignedURL.URL)
			return nil
		})

		g.Go(func() error {
			avStatus, err := s3Client.TagValueForKey(document.S3Key, upload.AVStatusTagName)
			if err != nil {
				return err
			}
			var status models.TRBRequestDocumentStatus

			if avStatus == "CLEAN" {
				status = models.TRBRequestDocumentStatusAvailable
			} else if avStatus == "INFECTED" {
				status = models.TRBRequestDocumentStatusUnavailable
			} else {
				status = models.TRBRequestDocumentStatusPending
			}
			documentStatuses.Store(document.ID, status)
			return nil
		})
	}

	if err := g.Wait(); err != nil {
		return nil, err
	}

  // load data from maps back into documents
	for _, document := range documents {
		if url, found := presignedURLs.Load(document.ID); found {
			document.URL = url.(string)
		}

		if status, found := documentStatuses.Load(document.ID); found {
			document.Status = status.(models.TRBRequestDocumentStatus)
		}
	}

	return documents, nil
}
```

Given that, I'm inclined to drop the URL field from the model and use it as a separate resolver. I _think_ gqlgen might parallelize multiple resolvers for the same query when possible; even if it doesn't, querying S3 to get the presigned URL is pretty quick, I don't think it's a significant performance issue.

Reply:
https://github.com/CMSgov/easi-app/pull/1819/commits/405f2249d2209d55029e14450e72f147721468b8 separates URL out into its own resolver; I did the same thing for Status in https://github.com/CMSgov/easi-app/pull/1819/commits/b3170ccc370ffbd1cab7bfe186c18e04e68bd4ce.